lib/collection/src/shard/forward_proxy_shard.rs

```diff
@@ -0,0 +1,166 @@
+use std::path::Path;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use segment::types::{
+    ExtendedPointId, Filter, PointIdType, ScoredPoint, WithPayload, WithPayloadInterface,
+};
+use tokio::runtime::Handle;
+use tokio::sync::Mutex;
+
+use crate::operations::point_ops::{PointInsertOperations, PointOperations};
+use crate::operations::types::{
+    CollectionInfo, CollectionResult, CountRequest, CountResult, PointRequest, Record,
+    SearchRequest, UpdateResult,
+};
+use crate::operations::CollectionUpdateOperations;
+use crate::shard::local_shard::LocalShard;
+use crate::shard::remote_shard::RemoteShard;
+use crate::shard::ShardOperation;
+
+/// ForwardProxyShard
+///
+/// ForwardProxyShard is a wrapper type for a LocalShard.
+///
+/// It can be used to provide all read and write operations while the wrapped shard is being transferred to another node.
+/// Proxy forwards all operations to remote shards.
+pub struct ForwardProxyShard {
+    wrapped_shard: LocalShard,
+    remote_shard: RemoteShard,
+    /// Lock required to protect transfer-in-progress updates.
+    /// It should block data updating operations while the batch it being transferred.
+    update_lock: Mutex<()>,
+}
+
+impl ForwardProxyShard {
+    pub fn new(wrapped_shard: LocalShard, remote_shard: RemoteShard) -> Self {
+        Self {
+            wrapped_shard,
+            remote_shard,
+            update_lock: Mutex::new(()),
+        }
+    }
+
+    /// Move batch of points to the remote shard.
+    /// Returns an offset of the next batch to be transferred.
+    pub async fn transfer_batch(
+        &self,
+        offset: Option<PointIdType>,
+        batch_size: usize,
+    ) -> CollectionResult<Option<PointIdType>> {
+        debug_assert!(batch_size > 0);
+        let limit = batch_size + 1;
+        let _update_lock = self.update_lock.lock().await;
+        let mut batch = self
+            .wrapped_shard
+            .scroll_by(offset, limit, &WithPayloadInterface::Bool(true), true, None)
+            .await?;
+        let next_page_offset = if batch.len() < limit {
+            // This was the last page
+            None
+        } else {
+            // remove extra point, it would be a first point of the next page
+            Some(batch.pop().unwrap().id)
+        };
+
+        if batch.is_empty() {
+            return Ok(next_page_offset);
+        }
+
+        let insert_points_operation = CollectionUpdateOperations::PointOperation(
+            PointOperations::UpsertPoints(PointInsertOperations::PointsList(
+                batch.into_iter().map(|point| point.into()).collect(),
+            )),
+        );
+
+        // We only need to wait for the last batch.
```

Comment:
Why do we wait only for the last batch?
My first impression is that it would be nicer for the receiving node to respect back-pressure all the way.